mindmeld/auto_annotator.py

```diff
@@ -835,5 +839,300 @@ def valid_entity_check(self, entity):
+        duckling_entities = (
+            self._parse_with_translator(**params)
+            if self.translator
+            else self._parse_without_translator(**params)
+        )
+        if entity_types:
+            duckling_entities = [
+                e for e in duckling_entities if e["dim"] in entity_types
+            ]
+        return MultiLingualAnnotator.convert_duckling_candidates(duckling_entities)
+
+    def _parse_with_translator(self, sentence, language=None):
+        """Parse helper function to be used if a translator is present.
+
+        Args:
+            sentence (str): Sentence to detect entities.
+            language (str): Language code for the language of the given sentence
+        Returns: entities (list): List of duckling candidates.
+        """
+        candidates = self.en_annotator.duckling.get_candidates_for_text(
+            sentence, language=language
+        )
+        en_sentence = self.translator.translate(sentence, target_language="en")
+        en_entities = self.en_annotator.parse(en_sentence)
+
+        selected_candidates = []
+        for entity in en_entities:
+            i = 0
+            while i < len(candidates):
+                if entity["dim"] == candidates[i]["entity_type"]:
+                    if entity["value"] == candidates[i]["value"]:
+                        selected_candidates.append(candidates[i])
+                        break
+                    if (
+                        self.translator.translate(
+                            entity["body"], target_language=language
+                        )
+                        == candidates[i]["body"]
+                    ):
+                        selected_candidates.append(candidates[i])
+                        break
+                i += 1
+        return selected_candidates
+
+    def _parse_without_translator(self, sentence, language=None):
+        """Parse helper function to be used if a translator is not present.
+
+        Args:
+            sentence (str): Sentence to detect entities.
+            language (str): Language code for the language of the given sentence
+        Returns: entities (list): List of duckling candidates.
+        """
+        candidates = self.en_annotator.duckling.get_candidates_for_text(
+            sentence, language=language
+        )
+        final_spans = MultiLingualAnnotator._get_final_spans(candidates, sentence)
+        selected_candidates = []
+        for span in final_spans:
+            for candidate in candidates:
+                if span == (candidate["start"], candidate["end"]):
+                    if not (
+                        candidate["dim"] == "amount-of-money"
+                        and candidate["value"]["unit"] == "unknown"
+                    ):
+                        selected_candidates.append(candidate)
+                        break
+        return selected_candidates
+
+    @staticmethod
+    def convert_duckling_candidates(duckling_candidates):
```

Comment:
This should be in the duckling class

Reply:
Moved to the system_entity_recognizer.py as duckling_item_to_entity_dict(). Placed in parallel to related functions: duckling_item_to_query_entity(), duckling_item_to_entity().